[PrintAsClang] Kill declaration sorting bugs super-dead

beccadax opened the pull request:
PrintAsClang is supposed to emit declarations in the same order regardless of the compiler’s internal state, but we have repeatedly found that our current criteria are inadequate, resulting in non-functionality-affecting changes to generated header content. This PR is an attempt to address this problem more comprehensively:

* Corrects a bug that stopped several extension sorting rules from taking effect.
* Adds several new rules, like interface type and generic signature, to sort extensions by members.
* Adds a new rule that sorts an extension as a whole by its generic signature.
* Most importantly, emits diagnostics if PrintAsClang is unable to stably sort two declarations.
* Tests those diagnostics by enabling `-verify` on a large number of existing tests, demonstrating that those diagnostics are not emitted.

The new diagnostics—which users *should* never see, though I wouldn't be surprised to learn that there are still edge cases to fix—look like this:

```
Swift._RuntimeFunctionCountersStats:1:11: warning: arbitrarily printing extension of '_RuntimeFunctionCountersStats' before extension of '_RuntimeFunctionCountersStats' in generated header; subsequent compilations may print them in a different order
1 | extension _RuntimeFunctionCountersStats {
  |           |- warning: arbitrarily printing extension of '_RuntimeFunctionCountersStats' before extension of '_RuntimeFunctionCountersStats' in generated header; subsequent compilations may print them in a different order
  |           `- note: please report this as a bug in the Swift compiler and describe both declarations
2 |     public func dump(skipUnchanged: Bool)
3 |     public func dumpDiff(_ after: Self, skipUnchanged: Bool)

Swift._RuntimeFunctionCountersStats:1:11: note: extension of '_RuntimeFunctionCountersStats' will be printed after extension of '_RuntimeFunctionCountersStats' during this compilation
1 | extension _RuntimeFunctionCountersStats {
  |           `- note: extension of '_RuntimeFunctionCountersStats' will be printed after extension of '_RuntimeFunctionCountersStats' during this compilation
2 |     public var debugDescription: String { get }
3 | }
```

Fixes rdar://143533893 (harder).

beccadax committed:
[PrintAsClang] Beef up extension sorting rule

A couple of the rules that `ModuleContentsWriter::write()` uses to sort declarations didn’t actually work because of an incorrect predicate. In addition, there were a number of situations that could come up in C++ interop (where overloading is permitted) where extensions could not be sorted. Rework extension sorting to look for more kinds of differences between extension members.

beccadax committed:
[PrintAsClang] Warn about unstable decl order

PrintAsClang is supposed to emit declarations in the same order regardless of the compiler’s internal state, but we have repeatedly found that our current criteria are inadequate, resulting in non-functionality-affecting changes to generated header content. Add a diagnostic that’s emitted when this happens soliciting a bug report.

Since there *should* be no cases where the compiler fails to order declarations, this diagnostic is never actually emitted. Instead, we test this change by enabling `-verify` on nearly all PrintAsClang tests to make sure they are unaffected.

This did demonstrate a missing criterion that only mattered in C++ mode: extensions that varied only in their generic signature were not sorted stably. Add a sort criterion for this.

beccadax commented:
@swift-ci please test

beccadax commented:
@swift-ci please test

beccadax commented:
@swift-ci please test

Xazax-hun commented on lib/PrintAsClang/ModuleContentsWriter.cpp:
```diff
@@ -780,44 +804,65 @@ class ModuleWriter {
+
+          auto lhsGeneric = lhsMember->getAsGenericContext(),
+               rhsGeneric = rhsMember->getAsGenericContext();
+          if (lhsGeneric && rhsGeneric) {
```
Should we handle the case when one of the ptrs are `null`?
Like:
```
          if (!rhsGeneric && lhsGeneric)
             return Descending;
           if (!lhsGeneric && rhsGeneric)
             return Ascending;
```

beccadax replied:
I've addressed this in the refactoring: the `getGenericSignatureString()` helper returns an empty string, so it gets factored into the comparison.

Xazax-hun commented on lib/PrintAsClang/ModuleContentsWriter.cpp:
```diff
@@ -780,44 +804,65 @@ class ModuleWriter {
-            cast<ExtensionDecl>(*lhs)->getAllMembers().end()) {
-          auto *lhsMember = dyn_cast<ValueDecl>(*mismatch.first),
-               *rhsMember = dyn_cast<ValueDecl>(*mismatch.second);
+        for (auto pair : llvm::zip_equal(lhsMembers, rhsMembers)) {
```
This function starts to be quite big.  I wonder if we should factor out some of these comparisons into separate helper functions.

beccadax replied:
Pushed a pretty comprehensive refactoring just now.

Xazax-hun commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -6345,6 +6345,17 @@ ERROR(invalid_objc_swift_rooted_class,none,
 NOTE(invalid_objc_swift_root_class_insert_nsobject,none,
      "inherit from 'NSObject' to silence this error", ())
 
+WARNING(objc_header_sorting_arbitrary,none,
```
Is it possible to add a test for this diagnostic or we have no known cases that would trigger this?

beccadax replied:
We have no known cases that would trigger this, so I'm instead testing that it isn't emitted by turning on `-verify` everywhere I can. 

(If you think it's really important, I could add a flag that deliberately breaks the sort for testing purposes.)

Xazax-hun reviewed (approved):
Mostly looks good to me, some nits inline.

beccadax committed:
[PrintAsClang] [NFC] Refactor decl sorting

Factor ModuleContentsWriter’s declaration sorting out into a separate helper function, and additionally rework that function into a series of abstract comparisons supported by various helper functions. This declutters the function and makes the high-level logic it implements much more clear, at the cost of hiding much of the control flow inside a macro.

This also makes a very small change to the handling of generic signature comparisons: declarations without a generic signature will be factored into comparisons by comparing an empty string.

beccadax committed:
[PrintAsClang] Tweak fallback member sort rules

Compare the names of all extension members first, before attempting weirder and more expensive comparisons like stringified type and mangled name. This gives us a sort order that’s a little more comprehensible to humans.

beccadax commented:
@swift-ci please test

Xazax-hun reviewed (approved):
Thanks for the refactoring, looks much cleaner! Some small nits inline.

beccadax commented:
@swift-ci please test

beccadax commented:
@swift-ci please test

beccadax commented:
@swift-ci please test Linux platform

beccadax commented:
@swift-ci please smoke test
